### PR TITLE
Question answering tlt

### PR DESCRIPTION
**Pull request by yzhang123**
Changed QAModel to be compatible with tlt:
* particularly added additional inference function

### CODE REVIEW

File: nemo/collections/nlp/models/question_answering/qa_model.py [hunk 1]
```diff
@@ -146,76 +149,135 @@ def validation_epoch_end(self, outputs):
...
-            n_best_size=self._cfg.test_ds.n_best_size,
-            max_answer_length=self._cfg.test_ds.max_answer_length,
-            version_2_with_negative=self._cfg.dataset.version_2_with_negative,
-            null_score_diff_threshold=self._cfg.test_ds.null_score_diff_threshold,
-            do_lower_case=self._cfg.dataset.do_lower_case,
-        )
+        return self.validation_epoch_end(outputs)
+
+    @torch.no_grad()
+    def inference(
```
[the remark this answers is not in the document]
Comment (VahidooX) [reply nemo/collections/nlp/models/question_answering/qa_model.py:174]: I think file should be enough

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: examples/nlp/question_answering/question_answering_squad.py [hunk 2]
```diff
@@ -31,9 +31,9 @@ def main(cfg: DictConfig) -> None:
     log_dir = exp_manager(trainer, cfg.get("exp_manager", None))
     infer_datasets = [cfg.model.validation_ds, cfg.model.test_ds]
     for infer_dataset in infer_datasets:
-        if infer_dataset.output_prediction_file is not None:
+        if infer_dataset.get('output_prediction_file', None) is not None:
```
[the remark this answers is not in the document]
Comment (tkornuta-nvidia) [reply examples/nlp/question_answering/question_answering_squad.py:34]: So generally I am for dropping all of those param checks in favor of Structured Configs with default values.

Thanks to this the user will have those params in one place, instead of going through different model/datasset methods trying to figure out where they are set...

But it is up to you whether you want to follow this approach.

File: nemo/collections/nlp/data/question_answering_squad/qa_dataset.py [hunk 3]
```diff
@@ -83,8 +83,8 @@ def __init__(
         self.version_2_with_negative = version_2_with_negative
         self.processor = SquadProcessor(data_file=data_file, mode=mode)
         self.mode = mode
-        if mode not in ["eval", "train", "test"]:
-            raise ValueError(f"mode should be either 'train', 'eval', or 'test' but got {mode}")
+        if mode not in ["val", "train", "test", "infer"]:
+            raise ValueError(f"mode should be either 'train', 'val', or 'test' but got {mode}")
```
[the remark this answers is not in the document]
Comment (tkornuta-nvidia) [reply nemo/collections/nlp/data/question_answering_squad/qa_dataset.py:87]: Why this name change? How many splits SQuaD provides? 

In TLT we evaluate the trained model on "test" split, and use "validation" split for validation during training. How are you using this "eval" split?

File: nemo/collections/nlp/data/question_answering_squad/qa_squad_processing.py [hunk 4]
```diff
@@ -255,7 +255,7 @@ class SquadProcessor(DataProcessor):
 
     Args:
         data_file: data file path
-        mode: "train"/"eval" for training/validation mode
+        mode: "train"/"eval" for training/inf mode
```
[the remark this answers is not in the document]
Comment (tkornuta-nvidia) [reply nemo/collections/nlp/data/question_answering_squad/qa_squad_processing.py:258]: training/validation/infer?

File: nemo/collections/nlp/models/question_answering/qa_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tkornuta-nvidia) [reply nemo/collections/nlp/models/question_answering/qa_model.py:174]: Actually I like the standardized "infer()" more, it was @okuchaiev who wanted specific names :)

So really, it is up to you.

**Review by tkornuta-nvidia — CHANGES_REQUESTED** [review]
@VahidooX left plenty of good comments.

My main concern is the "eval" - it would be great if we would agree on meaning/role of this split and stick to it in all our examples.

I would go with `validation` for the splits that are used for validation during training.

File: nemo/collections/nlp/models/question_answering/qa_model.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/models/question_answering/qa_model.py:174]: I guess i will leave it at this then

File: examples/nlp/question_answering/question_answering_squad.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/question_answering/question_answering_squad.py:34]: done

File: examples/nlp/question_answering/question_answering_squad.py [hunk 5]
```diff
@@ -31,9 +31,9 @@ def main(cfg: DictConfig) -> None:
...
             infer_dataset.output_prediction_file = os.path.join(log_dir, infer_dataset.output_prediction_file)
-        if infer_dataset.output_nbest_file is not None:
+        if infer_dataset.get('output_nbest_file', None) is not None:
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/question_answering/question_answering_squad.py:36]: done

File: examples/nlp/question_answering/conf/question_answering_squad_config.yaml [hunk 6]
```diff
@@ -34,54 +34,41 @@ model:
...
+    # on one another.
+    null_score_diff_threshold: 0.0
+    # If null_score - best_non_null is greater than the threshold predict null.
+    n_best_size: 20
+    # The total number of n-best predictions to generate at testing.
     use_cache: true
     # if true does lower case
     do_lower_case: true
 
+    num_workers:  2
```
[the remark this answers is not in the document]
Comment (yzhang123) [reply examples/nlp/question_answering/conf/question_answering_squad_config.yaml:50]: done

File: nemo/collections/nlp/data/question_answering_squad/qa_dataset.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/data/question_answering_squad/qa_dataset.py:87]: I renamed it back to eval. These modes are only used for creating dataset + datalayer for training/evaluation/inference.
Inference dataset e.g. does not need labels.

File: nemo/collections/nlp/data/question_answering_squad/qa_squad_processing.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yzhang123) [reply nemo/collections/nlp/data/question_answering_squad/qa_squad_processing.py:258]: i changed it to training/evaluation/inference

**Comment by yzhang123** [comment]
> @VahidooX left plenty of good comments.
> 
> My main concern is the "eval" - it would be great if we would agree on meaning/role of this split and stick to it in all our examples.
> 
> I would go with `validation` for the splits that are used for validation during training.

Thanks @VahidooX @tkornuta-nvidia for your review. I addressed the "eval concern".  I created now TRAINING_MODE, EVALUATION_MODE and INFERENCE_MODE. This is not the same as the datasplit. The mode determines what dataset type to create, e.g. TRAINING_MODE is for training data and assumes only 1 label/answer. EVALUATION_MODE is used for both validation and test data and expects 1 or more possible labels. INFERENCE_MODE creates a dataset without labels.

By doing this SQuAD requires 1 Dataset for all modes, which I find cleaner. There is unfortunately no consistency between models. AFAIK text classification dataset switches between modes by checking whether a file (assumes labeled data) or queries (assumes unlabeled data) is passed.

For token classification dataset yet another approach is used, which is to use 2 different classes for labeled and unlabeled data respectively.

**Review by VahidooX — APPROVED** [review]

**Comment by tkornuta-nvidia** [comment]
> I created now TRAINING_MODE, EVALUATION_MODE and INFERENCE_MODE. This is not the same as the datasplit. The mode determines what dataset type to create, e.g. TRAINING_MODE is for training data and assumes only 1 label/answer. EVALUATION_MODE is used for both validation and test data and expects 1 or more possible labels. INFERENCE_MODE creates a dataset without labels.

Great, this makes perfectly sense to me, thanks 👍

**Comment by tkornuta-nvidia** [comment]
> By doing this SQuAD requires 1 Dataset for all modes, which I find cleaner. There is unfortunately no consistency between models. AFAIK text classification dataset switches between modes by checking whether a file (assumes labeled data) or queries (assumes unlabeled data) is passed.

I totally agree.

I have once concern though: why MODE? what mode means in this context? I honestly think in this case we are talking about different `splits`...

My understanding is:
 - you start with the `dataset` (contains all data)
 - you split dataset into smaller subsets (for training, validation, testing etc.) => hence `split`

Additionally, some datasets might contain `variants`, like e.g. in Visual Reasoning in CLEVR aside of the "main CLEVR variant" you have CoGenT-A and CoGenT-B variants with different combinations of object features... makes sense?

**Comment by tkornuta-nvidia** [comment]
> By doing this SQuAD requires 1 Dataset for all modes, which I find cleaner. There is unfortunately no consistency between models. AFAIK text classification dataset switches between modes by checking whether a file (assumes labeled data) or queries (assumes unlabeled data) is passed.
> 
> For token classification dataset yet another approach is used, which is to use 2 different classes for labeled and unlabeled data respectively.

Ok, maybe this is a good moment to bring it/clean it up a bit? Adding that as an "identified" issue in Iteration 6...
